IdentityCore/src/webview/MSIDWebviewAuthorization.m

```diff
@@ -26,34 +26,227 @@
+        @synchronized([MSIDWebviewAuthorization class])
+        {
+            s_currentWebSession = nil;
+        }
+    }
+}
+
+// Helper methods
++ (NSDictionary *)queryParametersFromURL:(NSURL *)url
+{
+    // Check for auth response
+    // Try both the URL and the fragment parameters:
+    NSDictionary *parameters = [url msidFragmentParameters];
+    if (parameters.count == 0)
+    {
+        parameters = [url msidQueryParameters];
+    }
+    return parameters;
+}
+
+
++ (MSIDWebOAuth2Response *)responseWithURL:(NSURL *)url
+                              requestState:(NSString *)requestState
+                             stateVerifier:(MSIDWebUIStateVerifier)stateVerifier
+                                   context:(id<MSIDRequestContext>)context
+                                     error:(NSError **)error
+{
+    // This error case *really* shouldn't occur. If we're seeing it it's almost certainly a developer bug
+    if ([NSString msidIsStringNilOrBlank:url.absoluteString])
+    {
+        if (error) {
+            *error = MSIDCreateError(MSIDOAuthErrorDomain, MSIDErrorNoAuthorizationResponse, @"No authorization response received from server.", nil, nil, nil, context.correlationId, nil);
+        }
+        return nil;
+    }
+    
+    NSDictionary *parameters = [self.class queryParametersFromURL:url];
+    
+    // Check if this is a WPJ response
+    MSIDWebWPJAuthResponse *wpjResponse = [[MSIDWebWPJAuthResponse alloc] initWithScheme:url.scheme
+                                                                              parameters:parameters
+                                                                                 context:context
+                                                                                   error:nil];
+    if (wpjResponse)
+    {
+        wpjResponse.url = url;
+        return wpjResponse;
+    }
+    MSID_LOG_INFO(context, @"This is not a WPJ response");
+    
+    // Check for AAD response,
+    NSError *aadError = nil;
+    MSIDWebAADAuthResponse *aadResponse = [[MSIDWebAADAuthResponse alloc] initWithParameters:parameters
+                                                                                requestState:requestState
+                                                                               stateVerifier:stateVerifier
+                                                                                     context:context
+                                                                                       error:&aadError];
+    if (aadResponse)
+    {
+        aadResponse.url = url;
+        return aadResponse;
+    }
+    
+    if (aadError)
+    {
+        if (error) *error = aadError;
+        return nil;
+    }
+    
+    MSID_LOG_INFO(context, @"This is not an AAD response");
+    
+    // It is then, a standard OAuth2 response
+    //
+    // For now, there is no logic to really land here. As there is no definitive condition for response
+    // not being a AAD response.
```

Comment:
Consider moving response creation to the factory and create AAD response only in the AAD factory and base response in the base factory